packages/library-base/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt

```diff
@@ -152,4 +153,18 @@ internal class SuspendableNotifier(
     fun unregisterCallbacks() {
         realm.unregisterCallbacks()
     }
+
+    /**
+     * Manually force a refresh of the Realm, moving it to the latest version.
+     * This will also trigger the evaluation of all change listeners, which will
+     * be triggered as normal if anything changed.
+     *
+     * @return a frozen reference to the version of the Realm after the refresh.
+     */
+    suspend fun refresh(): FrozenRealmReference {
+        return withContext(dispatcher) {
+            RealmInterop.realm_refresh(realm.realmReference.dbPointer)
```

Comment:
Hmm, no. But the `realm` reference here is a Live Reference, and it is the one we refresh and get the snapshot from, so it should be the latest version, no?

This is also the reason for my comment in RealmImpl. `NotifierRealm.onRealmChanged()` will be triggered as well _eventually_  and there is a race condition with regards to who gets to update the RealmImpl first.

But if I understand things correctly, that should be fine?

Reply:
The `snapshot` is currently only updated on `onRealmChanged` (actually it is cleared in `onRealmChanged` and lazily refreshed when needed). The reason for this was that I didn't want to create new frozen reference each time we requested a snapshot. We are potentially issuing many different object notifications from different C-API callbacks on the same versions, so I tried to reuse the same frozen reference across all notifications from the same version (our logic/clean-up-strategy might also break if we have multiple `RealmReference` objects pointing to the same underlying native shared realm 🤔). By only updating it on the `onRealmChanged` callback we ensure that we only have one reference per version (and per the guarantee that `onRealmChanged` is issued before any other callback it should be safe to use that version for any other notification). And by only creating the snapshot when needed we ensure that we don't bother bookkeeping references for when there is no object notification listeners (since we are now using new references to `RealmImpl` on each `onRealmChanged` this optimization is maybe worthless). 

Since this code is running in the single threaded dispatcher you could maybe just clear the `snapshot` after the refresh, but could it be that the `onRealmChanged` was actually guaranteed to be triggered synchronously when already on the right thread? If not, then we should maybe do a pattern around this that ensures we remember to wipe the `snapshot` on each explicit update like this in the future.

Reply:
Hmm, sounds like I need to think a bit more about this implementation.

Reply:
Instead of calling `realm.snapshot` we could call `realmChanged().first()` as a way to hook into the change mechanism, but then we probably run into a race condition where we don't know if a change event is a buffered one (and thus older than the one we need).

I suspect we are forced to do some kind of synchronous thing here, and then verify that cleanup logic doesn't break if we have multiple versions of the same Realm 🤔

Reply:
I think I found a way to hook into the notification mechanism so we still only have one frozen instance...It does introduce a race to update `RealmImpl`, but since RealmImpl already supports this and I don't expect these API's to be called frequently, that is probably fine.

Reply:
Also see https://mongodb.slack.com/archives/C010R3CMP3N/p1649411267690299